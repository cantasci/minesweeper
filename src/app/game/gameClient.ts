export const SOCKET_ENDPOINT = 'wss://hometask.eg1236.com/game1/';

export type Socket = WebSocket;
class GameClient {
  private static socketInstance: Socket;

  public static get socket() {
    return this.socketInstance;
  }

  public static set socket(socketConnection: Socket) {
    this.socketInstance = socketConnection;
  }

  public static createConnection(url: string = SOCKET_ENDPOINT) {
    if (GameClient.socket) {
      return GameClient.socket;
    }
    const socketConnection = new WebSocket(url);
    GameClient.socket = socketConnection;
    return GameClient.socket;
  }
}

export { GameClient };
