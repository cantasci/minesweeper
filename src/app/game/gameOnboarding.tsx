import React, { FC } from 'react';

import { Typography } from '@mui/material';

const GameOnboarding: FC = () => {
  return <Typography fontSize={30}>Press START to play Minesweeper!</Typography>;
};

export default GameOnboarding;
