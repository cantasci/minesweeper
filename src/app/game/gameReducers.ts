import { createAction, createSlice, PayloadAction } from '@reduxjs/toolkit';

function convertMapPayload(payload: string): string[] {
  const rowList = payload.split('map:')[1].split('\n');
  return rowList.filter((item: string) => !!item.length);
}

interface GameState {
  map: string[];
  message: string;
  level: number;
}

const initialState: GameState = {
  map: [],
  message: '',
  level: 1,
};

export const initializeGame = createAction('startWatchingChannel');

const gameSlice = createSlice({
  name: 'game',
  initialState,
  reducers: {
    createNewGame(state) {
      state.level = 1;
      state.message = '';
      state.map = [];
    },
    setMap(state, action) {
      state.map = convertMapPayload(action.payload);
    },
    setLevel(state, action: PayloadAction<number>) {
      state.level = action.payload;
    },
    updateMessage(state, action) {
      state.message = action.payload;
    },
  },
});

export const { setMap, setLevel, updateMessage, createNewGame } = gameSlice.actions;

export const gameReducer = gameSlice.reducer;
