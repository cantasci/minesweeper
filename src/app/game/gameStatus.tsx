import React, { FC } from 'react';

import { styled } from '@mui/material';

const Status = styled('p')(() => ({
  fontSize: 20,
  fontWeight: 'bold',
  color: 'red',
}));

const GameStatus: FC<{ status?: string }> = ({ status }) => {
  return <Status>{status}</Status>;
};

export default GameStatus;
