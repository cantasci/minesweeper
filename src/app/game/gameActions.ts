import { eventChannel } from 'redux-saga';
import { take, put, call, apply, fork } from 'redux-saga/effects';

import { GameClient, Socket } from './gameClient';
import { createNewGame, setMap, updateMessage } from './gameReducers';

function createSocketChannel(socket: Socket) {
  return eventChannel((emit) => {
    const handleOnMessage = (event: MessageEvent) => {
      emit(event.data);
    };

    const errorHandler = (errorEvent: Event) => {
      emit(new Error(errorEvent?.type || 'UKNOWN'));
    };

    socket.addEventListener('message', handleOnMessage);
    socket.addEventListener('error', errorHandler);

    const unsubscribe = () => {
      socket.removeEventListener('message', handleOnMessage);
    };

    return unsubscribe;
  });
}

function* getMap(socket: Socket) {
  yield apply(socket, socket.send, ['map']);
}

export function createGame(payload: string) {
  GameClient.socket.send(payload);
}

export function* watchOnGame(): any {
  const socket = yield call(GameClient.createConnection);
  const socketChannel = yield call(createSocketChannel, socket);

  while (true) {
    try {
      const data = yield take(socketChannel);
      if (data.includes('map:')) {
        yield put(setMap(data));
      }
      if (data.includes('new:')) {
        yield put(createNewGame());
        yield fork(getMap, socket);
      }
      if (data.includes('open:')) {
        yield put(updateMessage(data.split('open: ')[1]));
        yield fork(getMap, socket);
      }
    } catch (err) {
      socketChannel.close();
    }
  }
}
