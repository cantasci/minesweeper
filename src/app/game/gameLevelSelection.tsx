import React from 'react';

import { FormControl, InputLabel, MenuItem, Select, SelectChangeEvent } from '@mui/material';
import { makeStyles } from '@mui/styles';

import { setLevel } from './gameReducers';

import { useAppDispatch, useAppSelector } from '../../store/hooks';

export const useLevelSelectionStyle = makeStyles({
  levelSelector: {
    maxWidth: '180px !important',
    height: 55,
  },
});

function GameLevelSelection() {
  const dispatch = useAppDispatch();
  const classes = useLevelSelectionStyle();
  const gameState = useAppSelector((state) => state.game);

  const handleOnLevelChange = (event: SelectChangeEvent) => {
    const newLevel = Number(event?.target?.value);
    dispatch(setLevel(newLevel));
  };

  const levelItems = Array(4)
    .fill(null)
    .map((_item, index) => index + 1);

  return (
    <FormControl fullWidth className={classes.levelSelector}>
      <InputLabel id="level-select-label">Level</InputLabel>
      <Select
        labelId="level-select-label"
        id="level-select"
        value={`${gameState.level}`}
        label="Level"
        onChange={handleOnLevelChange}
      >
        {levelItems.map((item) => (
          <MenuItem key={`level-item-${item}`} value={item}>
            {item}
          </MenuItem>
        ))}
      </Select>
    </FormControl>
  );
}

export default GameLevelSelection;
