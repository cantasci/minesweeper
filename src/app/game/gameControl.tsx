import React from 'react';

import { Button } from '@mui/material';
import { makeStyles } from '@mui/styles';

import { createGame } from './gameActions';
import GameLevelSelection from './gameLevelSelection';

import { useAppSelector } from '../../store/hooks';

export const useGameControlStyles = makeStyles({
  container: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
    marginTop: 20,
  },

  content: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    flexDirection: 'row',
    width: 300,
  },
  startButton: {
    minWidth: '100px !important',
    height: 55,
    marginLeft: 10,
  },
});

function GameControl() {
  const classes = useGameControlStyles();
  const gameState = useAppSelector((state) => state.game);

  const onPlayGame = () => {
    createGame(`new ${gameState.level}`);
  };

  return (
    <div className={classes.container}>
      <div className={classes.content}>
        <GameLevelSelection />
        <Button
          onClick={onPlayGame}
          variant="contained"
          color="success"
          className={classes.startButton}
          data-testid="start-game-btn"
        >
          {gameState.map.length ? 'Restart' : 'Start'}
        </Button>
      </div>
    </div>
  );
}

export default GameControl;
