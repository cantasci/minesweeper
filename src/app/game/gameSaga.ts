import { takeLatest } from 'redux-saga/effects';

import { watchOnGame } from './gameActions';
import { initializeGame } from './gameReducers';

export function* watcherGameSaga() {
  yield takeLatest(initializeGame.type, watchOnGame);
}
