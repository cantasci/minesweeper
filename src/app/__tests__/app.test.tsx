import React from 'react';

import { cleanup, render } from '@testing-library/react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import createSagaMiddleware from 'redux-saga';

import App from '../app';
import { watcherGameSaga } from '../game/gameSaga';

const sagaMiddleware = createSagaMiddleware();
const mockStore = configureStore([sagaMiddleware]);

describe('<App />', () => {
  afterEach(() => {
    jest.clearAllMocks();
    cleanup();
  });
  it('renders correctly and to match snapshot', () => {
    const initialState = {
      game: {
        map: [],
        message: '',
        level: 1,
      },
    };
    const store = mockStore(initialState);
    const tree = renderer
      .create(
        <Provider store={store}>
          <App />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('Start button appears in first render', () => {
    const initialState = {
      game: {
        map: [],
        message: '',
        level: 1,
      },
    };
    const store = mockStore(initialState);
    const wrapper = mount(
      <Provider store={store}>
        <App />
      </Provider>
    );
    expect(wrapper.text().includes('Minesweeper')).toBe(true);
    expect(wrapper.text().includes('Press START to play Minesweeper!')).toBe(true);
    expect(wrapper.text().includes('Start')).toBe(true);
  });
  it('Start button appears in first render', async () => {
    const initialState = {
      game: {
        map: [
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
        ],
        message: '',
        level: 1,
      },
    };
    const store = mockStore(initialState);
    sagaMiddleware.run(watcherGameSaga);

    const wrapper = render(
      <Provider store={store}>
        <App />
      </Provider>
    );

    expect(wrapper.getByText('Restart')).toBeTruthy();
  });
});
