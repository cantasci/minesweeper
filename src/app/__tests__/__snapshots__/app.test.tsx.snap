// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly and to match snapshot 1`] = `
<div
  className="css-1c3w8ki"
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-6z4eq8-MuiPaper-root"
  >
    <div
      className="css-0"
    >
      <p
        className="css-invog6"
      >
        Minesweeper
      </p>
    </div>
    <p
      className="css-1dij4kc"
    >
      
    </p>
    <p
      className="MuiTypography-root MuiTypography-body1 css-f8nw7v-MuiTypography-root"
    >
      Press START to play Minesweeper!
    </p>
    <div
      className="makeStyles-container-2"
    >
      <div
        className="makeStyles-content-3"
      >
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth makeStyles-levelSelector-5 css-q8hpuo-MuiFormControl-root"
        >
          <label
            className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={true}
            id="level-select-label"
          >
            Level
          </label>
          <div
            className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
            onClick={[Function]}
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="level-select-label level-select"
              className="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id="level-select"
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="button"
              tabIndex={0}
            >
              1
            </div>
            <input
              aria-hidden={true}
              className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              disabled={false}
              onAnimationStart={[Function]}
              onChange={[Function]}
              required={false}
              tabIndex={-1}
              value="1"
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-1in441m"
              >
                <span>
                  Level
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root makeStyles-startButton-4 css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          data-testid="start-game-btn"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Start
        </button>
      </div>
    </div>
    <div
      className="makeStyles-content-1"
    />
  </div>
</div>
`;
